backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java

```diff
@@ -2251,54 +2261,148 @@ public void deleteLogVersionedInputs(Session session, Long repoId, String commit
+              .build();
+      throw StatusProto.toStatusRuntimeException(status);
+    }
+    Map<String, List<ArtifactEntity>> artifactEntityMap = experimentRunObj.getArtifactEntityMap();
+
+    List<ArtifactEntity> result =
+        (artifactEntityMap != null && artifactEntityMap.containsKey(ModelDBConstants.ARTIFACTS))
+            ? artifactEntityMap.get(ModelDBConstants.ARTIFACTS)
+            : Collections.emptyList();
+    return result.stream()
+        .filter(artifactEntity -> artifactEntity.getKey().equals(key))
+        .findFirst();
+  }
+
   @Override
   public Entry<String, String> getExperimentRunArtifactsS3PathAndMultipartUploadID(
       String experimentRunId, String key, long partNumber, S3KeyFunction initializeMultipart)
       throws ModelDBException {
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
-      ExperimentRunEntity experimentRunObj =
-          session.get(ExperimentRunEntity.class, experimentRunId);
-      if (experimentRunObj == null) {
-        LOGGER.warn(ModelDBMessages.EXP_RUN_NOT_FOUND_ERROR_MSG);
+      ArtifactEntity artifactEntity = getArtifactEntity(session, experimentRunId, key);
+      return getS3PathAndMultipartUploadId(
+          session, artifactEntity, partNumber != 0, key, initializeMultipart);
+    }
+  }
+
+  public ArtifactEntity getArtifactEntity(Session session, String experimentRunId, String key)
+      throws ModelDBException {
+    Optional<ArtifactEntity> artifactEntityOptional =
+        getExperimentRunArtifact(session, experimentRunId, key);
+    return artifactEntityOptional.orElseThrow(
+        () -> new ModelDBException("Can't find specified artifact", io.grpc.Status.Code.NOT_FOUND));
+  }
+
+  private SimpleEntry<String, String> getS3PathAndMultipartUploadId(
+      Session session,
+      ArtifactEntity artifactEntity,
+      boolean partNumberSpecified,
+      String key,
+      S3KeyFunction initializeMultipart) {
+    String uploadId;
+    if (partNumberSpecified) {
+      uploadId = artifactEntity.getUploadId();
+      try {
+        String message = null;
+        if (uploadId == null) {
+          if (initializeMultipart == null) {
+            message = "Multipart wasn't initialized";
+          } else {
+            uploadId = initializeMultipart.apply(artifactEntity.getPath()).orElse(null);
+          }
+        }
+        if (message != null) {
+          LOGGER.info(message);
+          throw new ModelDBException(message, io.grpc.Status.Code.FAILED_PRECONDITION);
+        }
+      } catch (ModelDBException e) {
         Status status =
             Status.newBuilder()
-                .setCode(Code.NOT_FOUND_VALUE)
-                .setMessage(ModelDBMessages.EXP_RUN_NOT_FOUND_ERROR_MSG)
+                .setCode(Code.INVALID_ARGUMENT_VALUE)
+                .setMessage(e.getMessage())
+                .addDetails(Any.pack(CreateJob.Response.getDefaultInstance()))
```

Comment:
This seems wrong. There are many things that could cause an exception, especially since we're calling `initializeMultipart`. Just saying it's the user's fault seems wrong.

Reply:
I think I implemented some complex code here. I have made some refactoring because really this exception can happen only during multipart commit. So it can happen if 1. get multipart URL wasn't called before commit multipart. 2. get multipart URL was called for NFS.

So 1 looks like the user's fault, right? But should 2 have some other type of response?

Reply:
We should never hit (2), right? We just ignore the part info from the user and send in the flag back saying we don't support.

Reply:
I guess. But the flag is in GetUrlForArtifact. So it's still possible to call CommitMultipartArtifact after that.